runtime/lua/vscode-neovim/api.lua

```diff
@@ -1,133 +1,170 @@
+---            - leave_selection: (boolean) Whether to preserve the selected range, only valid when range or line_range is specified
+---            - callback: (function(err: string|nil, ret: any))
+---                        Optional callback function to handle the action result.
+---                        The first argument is the error message, and the second is the result.
+---                        If no callback is provided, any error message will be shown as a notification in VSCode.
+function M.action(name, opts)
+  vim.validate({
+    name = { name, "string" },
+    opts = { opts, "table", true },
+  })
+  opts = opts or {}
+  vim.validate({
+    ["opts.callback"] = { opts.callback, "f", true },
+    ["opts.args"] = { opts.args, "t", true },
+    ["opts.range"] = { opts.range, "t", true },
+    ["opts.line_range"] = { opts.lien_range, "t", true },
+    ["opts.leave_selection"] = { opts.leave_selection, "b", true },
+  })
+  opts._ = 1
+  if opts.range and opts.line_range then
+    error([[Cannot specify both "range" and "line_range"]])
   end
-
-  local top_line = vim.fn.line("w0")
-  local diff = top_line - vscode_topline
-
-  if diff ~= 0 and (vscode_topline > 0) then
-    vim.fn.winrestview({
-      topline = vscode_topline,
-    })
+  if opts.callback then
+    opts.callback = add_callback(opts.callback)
   end
+  vim.schedule(function()
+    vim.rpcnotify(vim.g.vscode_channel, "vscode-action", name, opts)
+  end)
 end
 
----Close windows
----@param wins number[]
-function M.close_windows(wins)
-  for _, win in ipairs(wins) do
-    pcall(vim.api.nvim_win_close, win, true)
+--- Run action, this function is synchronous and blocking
+---@param name string The action name, generally a vscode command
+---@param opts? table Optional options table, all fields are optional
+---            - args: (table) Optional arguments for the action
+---            - range: (table) Specific range for the action. In visual mode, this parameter is generally not required.
+---                     The format is consistent with the LSP Protocol.
+---            - line_range: (table) An array [start_line, end_line], which has the same effect as range. Lines are zero indexed
+---            - leave_selection: (boolean) Whether to preserve the selected range, only valid when range or line_range is specified
+---@param timeout? number Timeout in milliseconds. The default value is -1, which means no timeout.
+---
+---@return any: result
+function M.call(name, opts, timeout)
+  vim.validate({
+    name = { name, "string" },
+    opts = { opts, "table", true },
+    timeout = { timeout, "number", true },
+  })
+  opts = opts or {}
+  vim.validate({
+    ["opts.callback"] = { opts.callback, "nil" },
+    ["opts.args"] = { opts.args, "t", true },
+    ["opts.range"] = { opts.range, "t", true },
+    ["opts.line_range"] = { opts.line_range, "t", true },
+    ["opts.leave_selection"] = { opts.leave_selection, "b", true },
+  })
+  opts._ = 1
```

Comment:
What does this do?

Reply:
Empty table will be considered an array. I have modified it but haven’t pushed it yet.